source-jira-native/source_jira_native/api.py

```diff
@@ -0,0 +1,332 @@
+    log: Logger,
+) -> ZoneInfo:
+    url = f"{url_base(domain)}/myself"
+
+    response = MyselfResponse.model_validate_json(
+        await http.request(log, url)
+    )
+
+    timezone = ZoneInfo(response.timeZone)
+
+    log.info(f"Using timezone {response.timeZone} ({_format_utc_offset(timezone)}).")
+
+    return timezone
+
+
+async def snapshot_non_paginated_arrayed_resources(
+    http: HTTPSession,
+    domain: str,
+    path: str,
+    extra_params: dict[str, str] | None,
+    log: Logger,
+) -> AsyncGenerator[FullRefreshResource, None]:
+    url = f"{url_base(domain)}/{path}"
+
+    resources = TypeAdapter(list[FullRefreshResource]).validate_json(await http.request(log, url, params=extra_params))
+
+    for resource in resources:
+        yield resource
+
+
+async def snapshot_paginated_arrayed_resources(
+    http: HTTPSession,
+    domain: str,
+    path: str,
+    extra_params: dict[str, str] | None,
+    log: Logger,
+) -> AsyncGenerator[FullRefreshResource, None]:
+    url = f"{url_base(domain)}/{path}"
+
+    count = 0
+
+    params: dict[str, str | int] = {
+        "startAt": count,
+        # maxResults is more of a suggested maximum - Jira sometimes sends fewer 
+        # than maxResults results in a response depending on the size of the results.
+        "maxResults": 100,
+    }
+    if extra_params:
+        params.update(extra_params)
+
+    while True:
+        resources = TypeAdapter(list[FullRefreshResource]).validate_json(await http.request(log, url, params=params))
+
+        if len(resources) == 0:
+            break
+
+        for resource in resources:
+            yield resource
+            count += 1
+
+        params["startAt"] = count
+
+
+async def snapshot_paginated_resources(
+    http: HTTPSession,
+    domain: str,
+    path: str,
+    extra_params: dict[str, str] | None,
+    log: Logger,
+) -> AsyncGenerator[FullRefreshResource, None]:
+    url = f"{url_base(domain)}/{path}"
+
+    count = 0
+
+    params: dict[str, str | int] = {
+        "startAt": count,
+        # maxResults is more of a suggested maximum - Jira sometimes sends fewer 
+        # than maxResults results in a response depending on the size of the results.
+        "maxResults": 100,
+    }
+    if extra_params:
+        params.update(extra_params)
+
+    while True:
+        response = PaginatedResponse.model_validate_json(
+            await http.request(log, url, params=params)
+        )
+
+        for resource in response.values:
+            yield resource
+            count += 1
+
+        if response.isLast:
+            break
+
+        params["startAt"] = count
+
+
+def _build_jql(
+    start: datetime,
+    end: datetime,
+    projects: str | None,
+) -> str:
+    # There are a couple quirks with filtering issues by their updated field.
+    # - Top level timestamps are in the system default user timezone, not UTC.
+    # - Jira filters issues with minute-level granularity.
+    JQL_DATETIME_FORMAT = "%Y-%m-%d %H:%M"
```

Comment:
If updates occur within a minute, could a record be missed or duplicated between fetches?